scripts/install_python_dev_dependencies.py

```diff
@@ -41,13 +41,12 @@
     help='Assert that the dev requirements file is already compiled.')
 
 
-def assert_in_venv() -> None:
-    """Raise an error if we are not in a virtual environment.
-
-    No error is raised if we are running on GitHub Actions because a
-    virtual environment is unnecessary there.
-    """
+def check_python_env() -> None:
```

Comment:
I have only one real suggestion for this PR -- which is along the lines of my previous comment, to put the "if GITHUB_ACTION" in the caller. The reason is that the renaming of this function makes it ambiguous -- unlike before, I can no longer infer from the function name what it is supposed to do, and it hides a "special condition".

Is there a reason that the "if GITHUB_ACTION" check must be in this function?

Reply:
It doesn't have to be in this function, but I think the code is better decomposed when it is. In my mind, the point of this function is to check that the current python environment is suitable. That way, all the python-environment-related checks, which to me form a conceptual group, are together in a function.

To put it another way, if I break the operations this script performs into conceptual chunks, the chunks (steps) look like this:

1. Check that we are in a suitable environment--either in a virtual environment or on CI.
2. Install a minimal set of tools that we need to install the rest of the dependencies.
3. Make sure the `requirements_dev.txt` file is up-to-date with `requirements_dev.in` (requires the tools from step 2).
4. Install the development dependencies.

Right now, each step has its own function, which I think is easier to understand than having extra logic floating around in `main()`. It also makes the code more testable because the conceptually related code all gets tested in the same place as a unit. For example, currently we can test all the logic related to checking the environment in the tests for `check_python_env()`. If we moved the GITHUB_ACTION check to `main()` (which is the caller in this case), then the tests related to the python environment would be split between the tests for `main()` and those for `check_python_env()`, making it less obvious whether all the edge cases have been covered.

Of course in this case the code is simple enough that it would be fine to clutter up `main()` a little more, so I'm happy to move the GITHUB_ACTION check if you still think I should. I just wanted to take this opportunity to explain why I think it's beneficial to try and make functions map 1-to-1 with the conceptual units of work being done. That means we don't want a function to include code from two separate units (i.e. we don't want functions to be too big), but we also don't want a single unit of work to be split across multiple functions (i.e. we don't want functions to be too small).

Maybe a solution to satisfy both our concerns would be to just rename the function to more clearly describe what it does, e.g. `check_in_venv_or_on_ci()`. I also think `check_python_env_suitable()` is a reasonable level of abstraction for the `main()` function of a script, even though it's less explicit. It also better communicates how the function is supposed to be used (as opposed to just communicating what the function does).

Reply:
Ah, thanks for explaining the thinking behind it! That makes sense. I was thrown off because of the vague function name, and the way the docstring makes the github actions env look like an exception.

I think renaming the function to check_python_env_is_suitable() and changing the docstring to define "suitable" as your step 1 above (maybe with some explanation for why that's the right definition) would work fine, and alleviate my concern.

Reply:
done